rust/src/sdp/detect.rs

```diff
@@ -0,0 +1,1219 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// written by Giuseppe Longo <giuseppe@glongo.it>
+
+use crate::core::Direction;
+use crate::detect::{
+    DetectBufferSetActiveList, DetectHelperBufferMpmRegister, DetectHelperGetData,
+    DetectHelperGetMultiData, DetectHelperKeywordRegister, DetectHelperMultiBufferMpmRegister,
+    DetectSignatureSetAppProto, SCSigTableElmt, SIGMATCH_NOOPT,
+};
+use crate::sip::sip::{SIPTransaction, ALPROTO_SIP};
+use std::os::raw::{c_int, c_void};
+use std::ptr;
+
+static mut G_SDP_ORIGIN_BUFFER_ID: c_int = 0;
+static mut G_SDP_SESSION_NAME_BUFFER_ID: c_int = 0;
+static mut G_SDP_SESSION_INFO_BUFFER_ID: c_int = 0;
+static mut G_SDP_URI_BUFFER_ID: c_int = 0;
+static mut G_SDP_EMAIL_BUFFER_ID: c_int = 0;
+static mut G_SDP_PHONE_NUMBER_BUFFER_ID: c_int = 0;
+static mut G_SDP_CONNECTION_DATA_BUFFER_ID: c_int = 0;
+static mut G_SDP_BANDWIDTH_BUFFER_ID: c_int = 0;
+static mut G_SDP_TIME_BUFFER_ID: c_int = 0;
+static mut G_SDP_REPEAT_TIME_BUFFER_ID: c_int = 0;
+static mut G_SDP_TIMEZONE_BUFFER_ID: c_int = 0;
+static mut G_SDP_ENCRYPTION_KEY_BUFFER_ID: c_int = 0;
+static mut G_SDP_ATTRIBUTE_BUFFER_ID: c_int = 0;
+static mut G_SDP_MEDIA_DESC_MEDIA_BUFFER_ID: c_int = 0;
+static mut G_SDP_MEDIA_DESC_SESSION_INFO_BUFFER_ID: c_int = 0;
+static mut G_SDP_MEDIA_DESC_CONNECTION_DATA_BUFFER_ID: c_int = 0;
+static mut G_SDP_MEDIA_DESC_ENCRYPTION_KEY_BUFFER_ID: c_int = 0;
+
+unsafe extern "C" fn sdp_session_name_setup(
+    de: *mut c_void, s: *mut c_void, _raw: *const std::os::raw::c_char,
+) -> c_int {
+    if DetectSignatureSetAppProto(s, ALPROTO_SIP) != 0 {
+        return -1;
+    }
+    if DetectBufferSetActiveList(de, s, G_SDP_SESSION_NAME_BUFFER_ID) < 0 {
+        return -1;
+    }
+    return 0;
+}
+
+unsafe extern "C" fn sdp_session_name_get(
+    de: *mut c_void, transforms: *const c_void, flow: *const c_void, flow_flags: u8,
+    tx: *const c_void, list_id: c_int,
+) -> *mut c_void {
+    return DetectHelperGetData(
+        de,
+        transforms,
+        flow,
+        flow_flags,
+        tx,
+        list_id,
+        sdp_session_name_get_data,
+    );
+}
+
+unsafe extern "C" fn sdp_session_name_get_data(
+    tx: *const c_void, direction: u8, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> bool {
+    let tx = cast_pointer!(tx, SIPTransaction);
+    let sdp_message = match direction.into() {
+        Direction::ToServer => tx.request.as_ref().and_then(|req| req.body.as_ref()),
+        Direction::ToClient => tx.response.as_ref().and_then(|resp| resp.body.as_ref()),
+    };
+    if let Some(sdp) = sdp_message {
+        let session_name = &sdp.session_name;
+        if !session_name.is_empty() {
```

Comment:
should we return an empty buffer so we can match on `sdp.session_name; bsize:0;`?

Reply:
This field is mandatory, the parser will fail if it's not found.

Reply:
I think this code should suggest it can be empty then.

Reply:
Should a rule like `sdp.session_info; bsize:0` trigger an alert if the `session_info` field is not present at all?

Reply:
No, cf https://github.com/OISF/suricata/pull/11906 which adds `absent` keyword to match when the field is not present